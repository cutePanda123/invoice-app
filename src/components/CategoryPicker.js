import React from 'react';
import PropTypes from 'prop-types';
import Ionicon from 'react-ionicons';

class CategoryPicker extends React.Component {
    constructor(props) {
        super(props);
    }

    selectCategory = (event, category) => {
        this.props.onSelectCategory(category);
        event.preventDefault();
    }

    render() {
        const { categories, selectedCategory } = this.props;
        const selectedCategoryId = selectedCategory && selectedCategory.id;
        return (
            <div className="category-picker-component">
                <div className="row">
                    {
                        categories.map((category, index) => {
                            const iconColor = (category.id === selectedCategoryId) ? '#fff' : '#555';
                            const backColor = (category.id === selectedCategoryId) ? '#347eff' : '#efefef';
                            const activeClassName = selectedCategoryId == category.id ? 
                                "category-item col-3 active" : "category-item col-3";
                            return (
                                <div
                                    className={activeClassName}
                                    key={index}
                                    onClick={(event) => {
                                        this.selectCategory(event, category);
                                    }}
                                    role='button'
                                    style={{textAlign: 'center'}}
                                >
                                    <Ionicon 
                                        className="rounded-circle"
                                        style={
                                            {
                                                backgroundColor: backColor,
                                                padding: '5px'
                                            }
                                        }
                                        fontSize="50px"
                                        color={iconColor}
                                        icon={category.iconName}
                                    />
                                    <p>{category.name}</p>
                                </div>
                            );
                        })
                    }
                </div>
            </div>
        );
    }
}

CategoryPicker.propTypes = {
    categories: PropTypes.array.isRequired,
    onSelectCategory: PropTypes.func.isRequired,
    selectedCategory: PropTypes.object
};

export default CategoryPicker;

