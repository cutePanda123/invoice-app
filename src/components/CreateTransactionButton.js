import React from 'react';
import Ionicon from 'react-ionicons';
import PropTypes from 'prop-types';

const CreateTransactionButton = ({ onCreateTransaction }) => {
    return (
        <button 
            className="btn btn-primary btn-block d-flex justify-content-center align-items-center"
            onClick={(e) => {
                onCreateTransaction()
            }}    
        >
            <Ionicon
                className="rounded-circle"
                fontSize="30px"
                color="#fff"
                icon="ios-add-circle"
            />
            Create a new transaction
        </button>
    );
}

CreateTransactionButton.propTypes = {
    onCreateTransaction: PropTypes.func.isRequired
}

CreateTransactionButton.defaultProps = {
    onCreateTransaction: () => {
        console.log("create a transaction button");
    }
}

export default CreateTransactionButton;