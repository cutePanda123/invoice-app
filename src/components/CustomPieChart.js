import React from 'react';
import PropTypes from 'prop-types';
import { ResponsiveContainer, PieChart, Pie, Cell, Tooltip } from 'recharts';
import Utility from '../utility';

const Colors = Object.keys(Utility.Colors).map(key => Utility.Colors[key]);

const CustomPieChart = ({title, categoryData}) => {
    if (categoryData.length === 0) {
        return (
            <h3 className="text-center mx-3">{title} has no data</h3>
        );
    }
    return (
        <div className="pie-chart-component">
            <h3 className="text-center mt-3">{title}</h3>
            <ResponsiveContainer width={'100%'} height={300}>
                <PieChart>
                  <Pie
                    dataKey="value"
                    isAnimationActive={false}
                    data={categoryData}
                    cx={'50%'}
                    cy={'50%'}
                    outerRadius={100}
                    fill={Utility.Colors.blue}
                    label
                  >
                    {
                      categoryData.map((entry, index) => {
                        return (
                          <Cell key={`cell-${index}`} fill={Colors[index % Colors.length]} />
                        );
                      })
                    }
                  </Pie>
                  <Tooltip />
                </PieChart>
            </ResponsiveContainer>
        </div>
    );
};

CustomPieChart.propTypes = {
    title: PropTypes.string.isRequired,
    categoryData: PropTypes.array.isRequired
};

export default CustomPieChart;