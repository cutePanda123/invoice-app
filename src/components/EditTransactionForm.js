import React from 'react';
import PropTyes from 'prop-types';

class CreateTransactionForm extends React.Component {

    constructor(props) {
        super(props);

        this.state = {
            errorMessage: ''
        };
    }

    submitForm = (event) => {
        const {transaction, onFormSubmit} = this.props;
        const amount = this.amountInput.value.trim() * 1;
        const date = this.dateInput.value.trim();
        const title = this.titleInput.value.trim();
        const isEditMode = transaction.desc;

        if (amount && date && title) {
            if (amount < 0) {
                this.setState({
                    errorMessage: 'dollar amount cannot be negative'
                });
            } else {
                onFormSubmit({...transaction, amount, date, desc: title}, isEditMode)
            }
        } else {
            this.setState({
                errorMessage: 'input filed cannot be empty'
            });
        }
        event.preventDefault();
    }

    render() {
        return (
            <form onSubmit={(event) => {this.submitForm(event)}} noValidate>
                <div className="form-group">
                    <label htmlFor="title">Title</label>
                    <input
                        type="text"
                        className="form-control"
                        id="title"
                        aria-describedby="titleHelp"
                        placeholder="Enter title"
                        ref={(inputValue) => {
                            this.titleInput = inputValue;
                        }}
                        defaultValue={this.props.transaction.desc}
                    />
                    <small id="titleHelp" className="form-text text-muted">Title for your transaction.</small>
                </div>
                <div className="form-group">
                    <label htmlFor="amount">Dollar Amount</label>
                    <input
                        type="number"
                        className="form-control"
                        id="amount"
                        placeholder="Dollar Amount"
                        defaultValue={this.props.transaction.amount}
                        ref={(inputValue) => {
                            this.amountInput = inputValue;
                        }}
                    />
                    <small id="amountHelp" className="form-text text-muted">Dollar amount for your transaction.</small>
                </div>
                <div className="form-group">
                    <label htmlFor="date">Date</label>
                    <input
                        type="date"
                        className="form-control"
                        id="date"
                        placeholder="Date"
                        defaultValue={this.props.transaction.date}
                        ref={(inputValue) => {
                            this.dateInput = inputValue;
                        }}
                    />
                    <small id="dateHelp" className="form-text text-muted">Date for your transaction.</small>
                </div> 
                <button
                    type="submit"   
                    className="btn btn-primary"
                >Submit</button>
                <button
                    className="btn btn-secondary"
                    onClick={this.props.onFormCancel}
                >Cancel</button>
                {
                    this.state.errorMessage !== '' && 
                    <div className="alert alert-danger mt-5" role="alert">
                        {this.state.errorMessage}
                    </div>
                }
            </form>
        );
    }
}

CreateTransactionForm.propTypes = {
    transaction: PropTyes.object.isRequired,
    onFormSubmit: PropTyes.func.isRequired,
    onFormCancel: PropTyes.func.isRequired
}

CreateTransactionForm.defaultProps = {
    transaction: {}
}


export default CreateTransactionForm;