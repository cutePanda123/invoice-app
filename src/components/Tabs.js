import React from 'react';
import PropTypes from 'prop-types';

export class Tabs extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            activeIndex: this.props.activeIndex
        };
    }

    clickTab = (event, index) => {
        event.preventDefault();
        this.setState({
            activeIndex: index
        });
        this.props.onTabChange(index);
    }

    render() {
        const { children } = this.props;
        const { activeIndex } = this.state;
        return (
            <ul className="nav nav-tabs nav-fill my-4">
                {
                    React.Children.map(children, (child, index) => {
                        const curClassName = activeIndex === index ? 'nav-link active' : 'nav-link';
                        return (
                            <li className="nav-item">
                                <a
                                    onClick={(event) => {this.clickTab(event, index);}}
                                    className={curClassName}
                                    href="#"
                                >
                                    {child}
                                </a>
                            </li>
                        );
                    })
                }
            </ul>
        );
    }
};

Tabs.propTypes = {
    activeIndex: PropTypes.number.isRequired,
    onTabChange: PropTypes.func.isRequired
};

export const Tab = ({children}) => {
    return <React.Fragment>{children}</React.Fragment>;
};