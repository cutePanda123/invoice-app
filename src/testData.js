export const testCategories = [
  {
    "name": "travel",
    "iconName": "ios-plane",
    "id": "1",
    "type": "outcome"
  },
  {
    "name": "food",
    "iconName": "ios-restaurant",
    "id": "2",
    "type": "outcome"
  },
  {
    "name": "shopping",
    "iconName": "ios-basket",
    "id": "3",
    "type": "outcome"
  },
  {
    "name": "electronic product",
    "iconName": "ios-phone-portrait",
    "id": "4",
    "type": "outcome"
  },
  {
    "name": "工资",
    "iconName": "ios-card",
    "id": "10",
    "type": "income"
  },
  {
    "name": "part-time job",
    "iconName": "ios-cash",
    "id": "11",
    "type": "income"
  },
  {
    "name": "investment",
    "iconName": "logo-yen",
    "id": "12",
    "type": "income"
  }
];
  
export const testItems = [
  {
    "id": 0,
    "desc": "travel in Seattle",
    "amount": 2000,
    "date": "2020-10-20",
    "categoryId": 1
  },
  {
    "id": 1,
    "desc": "travel in San Jose",
    "amount": 4000,
    "date": "2020-12-10",
    "categoryId": 1
  },
  {
    "id": 2,
    "desc": "stock investement",
    "amount": 14000,
    "date": "2020-12-15",
    "categoryId": 2
  }
];

export const fakeTransaction = {
  "id": 0,
  "desc": "fake new transaction",
  "amount": 14000,
  "date": "2020-12-10",
  "categoryId": 2
};