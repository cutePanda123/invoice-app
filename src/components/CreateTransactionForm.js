import React from 'react';
import PropTyes from 'prop-types';
import CreateTransactionButton from './CreateTransactionButton';

class CreateTransactionForm extends React.Component {

    constructor(props) {
        super(props);

        this.state = {
            errorMessage: ''
        };
    }

    submitForm = (event) => {

        event.preventDefault();
    }

    render() {
        return (
            <form onSubmit={(event) => {this.submitForm(event)}} noValidate>
                <div className="form-group">
                    <label for="title">Title</label>
                    <input
                        type="text"
                        className="form-control"
                        id="title"
                        aria-describedby="titleHelp"
                        placeholder="Enter title"
                        ref={(inputValue) => {
                            this.titleInput = inputValue;
                        }}
                    />
                    <small id="titleHelp" className="form-text text-muted">Title for your transaction.</small>
                </div>
                <div className="form-group">
                    <label for="price">Price</label>
                    <input
                        type="number"
                        className="form-control"
                        id="price"
                        placeholder="Price"
                        defaultValue={this.props.transaction.price}
                        ref={(inputValue) => {
                            this.priceInput = inputValue;
                        }}
                    />
                    <small id="priceHelp" className="form-text text-muted">Price for your transaction.</small>
                </div>
                <div className="form-group">
                    <label for="date">Date</label>
                    <input
                        type="date"
                        className="form-control"
                        id="date"
                        placeholder="Date"
                        defaultValue={this.props.transaction.date}
                        ref={(inputValue) => {
                            this.dateInput = inputValue;
                        }}
                    />
                    <small id="dateHelp" className="form-text text-muted">Date for your transaction.</small>
                </div> 
                <button
                    type="submit"   
                    className="btn btn-primary"
                >Submit</button>
                <button
                    className="btn btn-secondary"
                    onClick={this.props.onFormCancel}
                >Cancel</button>
                {
                    this.state.errorMessage !== '' && 
                    <div className="alert alert-danger mt-5" role="alert">
                        {this.state.errorMessage}
                    </div>
                }
            </form>
        );
    }
}


CreateTransactionForm.propTypes = {
    transaction: PropTyes.object.isRequired,
    onFormSubmit: PropTyes.func.isRequired,
    onFormCancel: PropTyes.func.isRequired
}

CreateTransactionForm.defaultProps = {
    transaction: {}
}


export default CreateTransactionForm;