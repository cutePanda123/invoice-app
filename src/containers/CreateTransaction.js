import React from 'react';
import CategoryPicker from '../components/CategoryPicker';
import { Tabs, Tab } from '../components/Tabs';
import EditTransactionForm from '../components/EditTransactionForm';
import withContext from '../WithContext';
import { withRouter } from 'react-router-dom';
import PropTypes from 'prop-types';
import Loader from '../components/Loader';

export class CreateTransaction extends React.Component {
    constructor(props) {
        super(props);
        const { items, categories } = this.props.data;
        let filteredCategories = Object.keys(this.props.data.categories).filter((cId) => {
            return this.props.data.categories[cId].type === this.getActiveSelectedCategoryType(0);
        }).map(cId => this.props.data.categories[cId]);
        const editTransaction = this.getEditTransaction();
        let tabIndex = 0;
        let curCategory = filteredCategories[0];
        if (editTransaction && editTransaction.id) {
            curCategory = categories[items[editTransaction.id].categoryId];
            tabIndex = curCategory.type === "income" ? 0 : 1;
        }
        this.state = {
            activeTabIndex: tabIndex,
            selectedCategory: curCategory
        };
    }

    componentDidMount() {
        const { transactionId } = this.props.match.params;
        this.props.actions.getEditTransactionData(transactionId).then(data => {
            const {modifiedTransaction, categories} = data;
            let filteredCategories = Object.keys(categories).filter((cId) => {
                return categories[cId].type === this.getActiveSelectedCategoryType(0);
            }).map(cId => categories[cId]);

            let tabIndex = 0;
            let curCategory = filteredCategories[0];
            if (transactionId) {
                curCategory = categories[modifiedTransaction.categoryId];
                tabIndex = curCategory.type === "income" ? 0 : 1;
            }
            this.setState({
                activeTabIndex: tabIndex,
                selectedCategory: curCategory
            });
        });
    }

    tabChange = (tabIndex) => {
        this.setState({
            activeTabIndex: tabIndex
        });
    }

    getActiveSelectedCategoryType = (tabIndex) => {
        return tabIndex === 0 ? "income" : "outcome";
    }

    cancelSubmit = () => {
        this.props.history.push('/');
    }

    finishSubmit= (formData, isEditMode) => {
        if (!isEditMode) {
            this.props.actions.createTransaction({...formData, categoryId: this.state.selectedCategory.id}).then(() => {
                this.props.history.push('/');
            });
        } else {
            this.props.actions.updateTransaction({...formData, categoryId: this.state.selectedCategory.id}).then(() => {
                this.props.history.push('/');
            });
        }
    }

    selectCategory = (category) => {
        this.setState({
            selectedCategory: category
        });
    }

    getEditTransaction = () => {
        const { items } = this.props.data;
        const {transactionId} = this.props.match.params;
        const editTransaction = (transactionId && items[transactionId]) ? items[transactionId] : {}
        return editTransaction;
    }

    render() {
        const { categories, isLoading } = this.props.data;
        let filteredCategories = Object.keys(categories).filter((cId) => {
            return categories[cId].type === this.getActiveSelectedCategoryType(this.state.activeTabIndex);
        }).map(cId => categories[cId]);
        return (
            <div
                className='create-transaction py-3 px-3 rounded mt-3'
                style={{background: '#fff'}}
            >
                {
                    isLoading &&
                    <Loader />
                }
                { !isLoading &&
                    <React.Fragment>
                        <Tabs
                            activeIndex={this.state.activeTabIndex}
                            onTabChange={this.tabChange}
                        >
                            <Tab>Income</Tab>
                            <Tab>Outcome</Tab>
                        </Tabs>
                        <CategoryPicker
                            categories={filteredCategories}
                            onSelectCategory={this.selectCategory}
                            selectedCategory={this.state.selectedCategory}
                        />
                        <EditTransactionForm
                            onFormSubmit={this.finishSubmit}
                            onFormCancel={this.cancelSubmit}
                            transaction={this.getEditTransaction()}
                        />
                    </React.Fragment>
                }
            </div>
        );
    }
};

CreateTransaction.propTypes = {
    data: PropTypes.object.isRequired,
    actions: PropTypes.object.isRequired,
    history: PropTypes.object.isRequired,
    match: PropTypes.object.isRequired
}

export default withRouter(withContext(CreateTransaction));