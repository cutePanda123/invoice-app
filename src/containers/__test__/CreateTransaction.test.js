import React from 'react';
import { mount } from 'enzyme';
import { Tabs, Tab } from '../../components/Tabs';
import { CreateTransaction } from '../CreateTransaction';
import EditTransactionForm from '../../components/EditTransactionForm';
import CategoryPicker from '../../components/CategoryPicker';
import AppContext from '../../AppContext';
import { testItems, testCategories } from '../../testData';
import Utility from '../../utility';
import { BrowserRouter as Router } from 'react-router-dom';
import CreateTransactionButton from '../../components/CreateTransactionButton';

let wrapper = null;

const fullyLoadedData = {
    items: Utility.flattenArray(testItems),
    categories: Utility.flattenArray(testCategories),
    isLoading: false
};

const emptyData = {
    items: {},
    categories: {},
    isLoading: true
};

const matchWithTransaction = {
    params: {
        transactionId: testItems[0].id
    }
};

const emptyMatch = {
    params: {}
}

const mockActions = {
    getEditTransactionData: jest.fn(),
    createTransaction: jest.fn(),
    updateTransaction: jest.fn()
};



const mockHistory = {
    push: jest.fn()
};

describe('CreateTransaction test', () => {
    beforeEach(() => {
        mockActions.getEditTransactionData.mockReturnValue(Promise.resolve(
            {
                modifiedTransaction: testItems[0],
                categories: Utility.flattenArray(testCategories)
            }
        ));
        mockActions.createTransaction.mockReturnValue(Promise.resolve(''));
        mockActions.updateTransaction.mockReturnValue(Promise.resolve(''));
    });
    it('initial data: snapshot test', () => {
        mockActions.getEditTransactionData().then(data => {

        });
        wrapper = mount(
            <CreateTransaction
                data={emptyData}
                actions={mockActions}
                history={mockHistory}
                match={emptyMatch}
            />
        )
        expect(wrapper).toMatchSnapshot();        
    });

    it('loaded data: snapshot test', () => {
        mockActions.getEditTransactionData().then(data => {

        });
        wrapper = mount(
            <CreateTransaction
                data={fullyLoadedData}
                actions={mockActions}
                history={mockHistory}
                match={matchWithTransaction}
            />
        )
        expect(wrapper).toMatchSnapshot();        
    });

    // it('should render correct components', () => {
    //     expect(wrapper.find(Tabs).length).toEqual(1);
    //     expect(wrapper.find(Tab).length).toEqual(2);
    //     expect(wrapper.find(CategoryPicker).length).toEqual(1);
    //     expect(wrapper.find(EditTransactionForm).length).toEqual(1);
    // });
});


