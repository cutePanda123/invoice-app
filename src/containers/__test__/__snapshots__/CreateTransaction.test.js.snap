// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateTransaction test initial data: snapshot test 1`] = `
<CreateTransaction
  actions={
    Object {
      "createTransaction": [MockFunction],
      "getEditTransactionData": [MockFunction] {
        "calls": Array [
          Array [],
          Array [
            undefined,
          ],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": Promise {},
          },
          Object {
            "type": "return",
            "value": Promise {},
          },
        ],
      },
      "updateTransaction": [MockFunction],
    }
  }
  data={
    Object {
      "categories": Object {},
      "isLoading": true,
      "items": Object {},
    }
  }
  history={
    Object {
      "push": [MockFunction],
    }
  }
  match={
    Object {
      "params": Object {},
    }
  }
>
  <div
    className="create-transaction py-3 px-3 rounded mt-3"
    style={
      Object {
        "background": "#fff",
      }
    }
  >
    <Loader>
      <div
        className="loading-component text-center"
      >
        <Ionicon
          beat={false}
          color="#347eff"
          fontSize="40px"
          icon="ios-refresh"
          rotate={true}
          shake={false}
          style={Object {}}
        >
          <styled.svg
            beat={0}
            className=""
            fill="#347eff"
            height="40px"
            rotate={1}
            shake={0}
            style={Object {}}
            viewBox="0 0 1024 1024"
            width="40px"
          >
            <svg
              className="sc-bdVaJa blRSEt"
              fill="#347eff"
              height="40px"
              rotate={1}
              style={Object {}}
              viewBox="0 0 1024 1024"
              width="40px"
            >
              <path
                d="M512 768.2c-141.4 0-256-114.6-256-256.2s114.6-256.2 256-256.2v-87.8l192 128-192 111.4v-111.6c-119.2 0-216.2 97-216.2 216.2s97 216.2 216.2 216.2 216.2-96.2 216.2-216.2h39.8c0 142-114.6 256.2-256 256.2z"
              />
            </svg>
          </styled.svg>
        </Ionicon>
        <h5>
          Loading...
        </h5>
      </div>
    </Loader>
  </div>
</CreateTransaction>
`;

exports[`CreateTransaction test loaded data: snapshot test 1`] = `
<CreateTransaction
  actions={
    Object {
      "createTransaction": [MockFunction],
      "getEditTransactionData": [MockFunction] {
        "calls": Array [
          Array [],
          Array [
            0,
          ],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": Promise {},
          },
          Object {
            "type": "return",
            "value": Promise {},
          },
        ],
      },
      "updateTransaction": [MockFunction],
    }
  }
  data={
    Object {
      "categories": Object {
        "1": Object {
          "iconName": "ios-plane",
          "id": "1",
          "name": "travel",
          "type": "outcome",
        },
        "10": Object {
          "iconName": "ios-card",
          "id": "10",
          "name": "full-time job",
          "type": "income",
        },
        "11": Object {
          "iconName": "ios-cash",
          "id": "11",
          "name": "part-time job",
          "type": "income",
        },
        "12": Object {
          "iconName": "logo-yen",
          "id": "12",
          "name": "investment",
          "type": "income",
        },
        "2": Object {
          "iconName": "ios-restaurant",
          "id": "2",
          "name": "food",
          "type": "outcome",
        },
        "3": Object {
          "iconName": "ios-basket",
          "id": "3",
          "name": "shopping",
          "type": "outcome",
        },
        "4": Object {
          "iconName": "ios-phone-portrait",
          "id": "4",
          "name": "electronic product",
          "type": "outcome",
        },
      },
      "isLoading": false,
      "items": Object {
        "0": Object {
          "amount": 2000,
          "categoryId": 1,
          "date": "2020-10-20",
          "desc": "travel in Seattle",
          "id": 0,
        },
        "1": Object {
          "amount": 4000,
          "categoryId": 1,
          "date": "2020-12-10",
          "desc": "travel in San Jose",
          "id": 1,
        },
        "2": Object {
          "amount": 14000,
          "categoryId": 12,
          "date": "2020-12-15",
          "desc": "stock investement",
          "id": 2,
        },
      },
    }
  }
  history={
    Object {
      "push": [MockFunction],
    }
  }
  match={
    Object {
      "params": Object {
        "transactionId": 0,
      },
    }
  }
>
  <div
    className="create-transaction py-3 px-3 rounded mt-3"
    style={
      Object {
        "background": "#fff",
      }
    }
  >
    <Tabs
      activeIndex={0}
      onTabChange={[Function]}
    >
      <ul
        className="nav nav-tabs nav-fill my-4"
      >
        <li
          className="nav-item"
          key=".0"
        >
          <a
            className="nav-link active"
            href="#"
            onClick={[Function]}
          >
            <Tab>
              Income
            </Tab>
          </a>
        </li>
        <li
          className="nav-item"
          key=".1"
        >
          <a
            className="nav-link"
            href="#"
            onClick={[Function]}
          >
            <Tab>
              Outcome
            </Tab>
          </a>
        </li>
      </ul>
    </Tabs>
    <CategoryPicker
      categories={
        Array [
          Object {
            "iconName": "ios-card",
            "id": "10",
            "name": "full-time job",
            "type": "income",
          },
          Object {
            "iconName": "ios-cash",
            "id": "11",
            "name": "part-time job",
            "type": "income",
          },
          Object {
            "iconName": "logo-yen",
            "id": "12",
            "name": "investment",
            "type": "income",
          },
        ]
      }
      onSelectCategory={[Function]}
      selectedCategory={
        Object {
          "iconName": "ios-card",
          "id": "10",
          "name": "full-time job",
          "type": "income",
        }
      }
    >
      <div
        className="category-picker-component"
      >
        <div
          className="row"
        >
          <div
            className="category-item col-3 active"
            key="0"
            onClick={[Function]}
            role="button"
            style={
              Object {
                "textAlign": "center",
              }
            }
          >
            <Ionicon
              beat={false}
              className="rounded-circle"
              color="#fff"
              fontSize="50px"
              icon="ios-card"
              rotate={false}
              shake={false}
              style={
                Object {
                  "backgroundColor": "#347eff",
                  "padding": "5px",
                }
              }
            >
              <styled.svg
                beat={0}
                className="rounded-circle"
                fill="#fff"
                height="50px"
                rotate={0}
                shake={0}
                style={
                  Object {
                    "backgroundColor": "#347eff",
                    "padding": "5px",
                  }
                }
                viewBox="0 0 1024 1024"
                width="50px"
              >
                <svg
                  className="rounded-circle sc-bdVaJa fUuvxv"
                  fill="#fff"
                  height="50px"
                  rotate={0}
                  style={
                    Object {
                      "backgroundColor": "#347eff",
                      "padding": "5px",
                    }
                  }
                  viewBox="0 0 1024 1024"
                  width="50px"
                >
                  <path
                    d="M864 224h-704c-35.2 0-64 28.8-64 64v448c0 35.2 28.8 64 64 64h704c35.2 0 64-28.8 64-64v-448c0-35.2-28.8-64-64-64zM160 256h704c17.6 0 32 14.4 32 32v64h-768v-64c0-17.6 14.4-32 32-32zM864 768h-704c-17.6 0-32-14.4-32-32v-224h768v224c0 17.6-14.4 32-32 32z M192 640h64v32h-64v-32z M320 640h384v32h-384v-32z"
                  />
                </svg>
              </styled.svg>
            </Ionicon>
            <p>
              full-time job
            </p>
          </div>
          <div
            className="category-item col-3"
            key="1"
            onClick={[Function]}
            role="button"
            style={
              Object {
                "textAlign": "center",
              }
            }
          >
            <Ionicon
              beat={false}
              className="rounded-circle"
              color="#555"
              fontSize="50px"
              icon="ios-cash"
              rotate={false}
              shake={false}
              style={
                Object {
                  "backgroundColor": "#efefef",
                  "padding": "5px",
                }
              }
            >
              <styled.svg
                beat={0}
                className="rounded-circle"
                fill="#555"
                height="50px"
                rotate={0}
                shake={0}
                style={
                  Object {
                    "backgroundColor": "#efefef",
                    "padding": "5px",
                  }
                }
                viewBox="0 0 1024 1024"
                width="50px"
              >
                <svg
                  className="rounded-circle sc-bdVaJa fUuvxv"
                  fill="#555"
                  height="50px"
                  rotate={0}
                  style={
                    Object {
                      "backgroundColor": "#efefef",
                      "padding": "5px",
                    }
                  }
                  viewBox="0 0 1024 1024"
                  width="50px"
                >
                  <path
                    d="M80 736h864v32h-864v-32z M96 800h832v32h-832v-32z M512 336c-61.8 0-112 50.2-112 112s50.2 112 112 112 112-50.2 112-112-50.2-112-112-112z M64 192v512h896v-512h-896zM256 640h-128v-32h128v32zM256 288h-128v-32h128v32zM512 592c-79.4 0-144-64.6-144-144s64.6-144 144-144 144 64.6 144 144-64.6 144-144 144zM896 640h-128v-32h128v32zM896 288h-128v-32h128v32z"
                  />
                </svg>
              </styled.svg>
            </Ionicon>
            <p>
              part-time job
            </p>
          </div>
          <div
            className="category-item col-3"
            key="2"
            onClick={[Function]}
            role="button"
            style={
              Object {
                "textAlign": "center",
              }
            }
          >
            <Ionicon
              beat={false}
              className="rounded-circle"
              color="#555"
              fontSize="50px"
              icon="logo-yen"
              rotate={false}
              shake={false}
              style={
                Object {
                  "backgroundColor": "#efefef",
                  "padding": "5px",
                }
              }
            >
              <styled.svg
                beat={0}
                className="rounded-circle"
                fill="#555"
                height="50px"
                rotate={0}
                shake={0}
                style={
                  Object {
                    "backgroundColor": "#efefef",
                    "padding": "5px",
                  }
                }
                viewBox="0 0 1024 1024"
                width="50px"
              >
                <svg
                  className="rounded-circle sc-bdVaJa fUuvxv"
                  fill="#555"
                  height="50px"
                  rotate={0}
                  style={
                    Object {
                      "backgroundColor": "#efefef",
                      "padding": "5px",
                    }
                  }
                  viewBox="0 0 1024 1024"
                  width="50px"
                >
                  <path
                    d="M896 64h-160l-224 442.256-224-442.256h-160l224.736 416h-96.736v96h147.128l28.872 62v34h-176v96h176v192h160v-192h176v-96h-176v-34l29.782-62h146.218v-96h-96.578l224.578-416z"
                  />
                </svg>
              </styled.svg>
            </Ionicon>
            <p>
              investment
            </p>
          </div>
        </div>
      </div>
    </CategoryPicker>
    <CreateTransactionForm
      onFormCancel={[Function]}
      onFormSubmit={[Function]}
      transaction={Object {}}
    >
      <form
        noValidate={true}
        onSubmit={[Function]}
      >
        <div
          className="form-group"
        >
          <label
            htmlFor="title"
          >
            Title
          </label>
          <input
            aria-describedby="titleHelp"
            className="form-control"
            id="desc"
            placeholder="Enter title"
            type="text"
          />
          <small
            className="form-text text-muted"
            id="titleHelp"
          >
            Title for your transaction.
          </small>
        </div>
        <div
          className="form-group"
        >
          <label
            htmlFor="amount"
          >
            Dollar Amount
          </label>
          <input
            className="form-control"
            id="amount"
            placeholder="Dollar Amount"
            type="number"
          />
          <small
            className="form-text text-muted"
            id="amountHelp"
          >
            Dollar amount for your transaction.
          </small>
        </div>
        <div
          className="form-group"
        >
          <label
            htmlFor="date"
          >
            Date
          </label>
          <input
            className="form-control"
            id="date"
            placeholder="Date"
            type="date"
          />
          <small
            className="form-text text-muted"
            id="dateHelp"
          >
            Date for your transaction.
          </small>
        </div>
        <button
          className="btn btn-primary"
          type="submit"
        >
          Submit
        </button>
        <button
          className="btn btn-secondary"
          onClick={[Function]}
        >
          Cancel
        </button>
      </form>
    </CreateTransactionForm>
  </div>
</CreateTransaction>
`;

exports[`CreateTransaction test snapshot test 1`] = `
<CreateTransaction
  actions={
    Object {
      "createTransaction": [MockFunction],
      "getEditTransactionData": [MockFunction] {
        "calls": Array [
          Array [],
          Array [
            undefined,
          ],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": Promise {},
          },
          Object {
            "type": "return",
            "value": Promise {},
          },
        ],
      },
      "updateTransaction": [MockFunction],
    }
  }
  data={
    Object {
      "categories": Object {},
      "isLoading": true,
      "items": Object {},
    }
  }
  history={
    Object {
      "push": [MockFunction],
    }
  }
  match={
    Object {
      "params": Object {},
    }
  }
>
  <div
    className="create-transaction py-3 px-3 rounded mt-3"
    style={
      Object {
        "background": "#fff",
      }
    }
  >
    <Loader>
      <div
        className="loading-component text-center"
      >
        <Ionicon
          beat={false}
          color="#347eff"
          fontSize="40px"
          icon="ios-refresh"
          rotate={true}
          shake={false}
          style={Object {}}
        >
          <styled.svg
            beat={0}
            className=""
            fill="#347eff"
            height="40px"
            rotate={1}
            shake={0}
            style={Object {}}
            viewBox="0 0 1024 1024"
            width="40px"
          >
            <svg
              className="sc-bdVaJa blRSEt"
              fill="#347eff"
              height="40px"
              rotate={1}
              style={Object {}}
              viewBox="0 0 1024 1024"
              width="40px"
            >
              <path
                d="M512 768.2c-141.4 0-256-114.6-256-256.2s114.6-256.2 256-256.2v-87.8l192 128-192 111.4v-111.6c-119.2 0-216.2 97-216.2 216.2s97 216.2 216.2 216.2 216.2-96.2 216.2-216.2h39.8c0 142-114.6 256.2-256 256.2z"
              />
            </svg>
          </styled.svg>
        </Ionicon>
        <h5>
          Loading...
        </h5>
      </div>
    </Loader>
  </div>
</CreateTransaction>
`;
